Add tdat reader - resolves issue 15190

d-giles opened the pull request:
<!-- These comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/main/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/main/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
https://docs.astropy.org/en/latest/development/quickstart.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/git_edit_workflow_examples.html . -->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable subpackage(s). -->

<!-- READ THIS FOR MANUAL BACKPORT FROM A MAINTAINER:
Apply "skip-basebranch-check" label **before** you open the PR! -->

This pull request is adds functionality to read and write to a Table from the TDAT format as part of the Unified File Read/Write Interface and includes tests for basic functionality.

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

Fixes #15190

<!-- Optional opt-out -->

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

taldcroft commented:
@d-giles - thanks for the PR, this looks like a great addition!

This is a big PR so it might take a bit to find time for a proper review. In the meantime some things to address are:

- Tests are currently failing.
- There appear to be some coverage issues in the new code.
- Add a changelog entry.
- Add a What's New entry.

pllim commented:
I think you should also rebase to get rid of the merge commit in your PR commit history. Otherwise it might come back to bite you later since this PR might be open for a while.

```
git fetch upstream main
git rebase upstream/main
```

You will need to force push, so if you are unsure, make a backup branch first. Thanks!

d-giles commented:
In the following failure:
`FAILED ../../.tox/test-alldeps/lib/python3.12/site-packages/astropy/table/tests/test_mixin.py::test_io_ascii_write`

The test attempts to write a `QTable` to the `tdat` format, but cannot for several reasons. Per the specifications for the tdat format, it really shouldn't be supported since tdat can only accept tables with int, float, or string data. I could rewrite the function such that any `Quantity` or `object` like data is simply interpreted as string data, though I think it may be misleading to a user to even support the option as it wouldn't pass validation checks down the road. Would it be better to circumvent the test as it shouldn't actually be supported, or rewrite the function so it can accomplish the behavior in the case that a user, for some ill advised reason, wants to output a QTable to the tdat format?

d-giles commented:
pre-commit.ci autofix

taldcroft commented:
@d-giles - since TDAT format supports `unit`, there should be no problem with reading/writing a `QTable`. The numeric part of the `Quantity` goes in the data and the unit goes to the `field` definition.

taldcroft commented:
These changes get the tests to pass. Basically the problem was using attribute access like `col.name` instead of `col.info.name`. The latter is needed to support mixin columns like `Quantity` where the column attributes are stored in `info`.
```
commit cd06375e13 (HEAD -> refs/heads/pr/d-giles/16780)
Author: Tom Aldcroft <taldcroft@gmail.com>
Date:   Mon Aug 19 09:07:54 2024 -0400

    Use col.info for column attributes like name
    
    This is needed for compatibility with
    mixin columns like Quantity

diff --git a/astropy/io/ascii/tdat.py b/astropy/io/ascii/tdat.py
index c69988ad76..12dbb64022 100644
--- a/astropy/io/ascii/tdat.py
+++ b/astropy/io/ascii/tdat.py
@@ -309,7 +309,7 @@ class TdatHeader(basic.BasicHeader):
         ]:
             keywords[key] = self.table_meta.get(key)
 
-        indices = [col.name for col in self.cols if col.indices != []]
+        indices = [col.info.name for col in self.cols if col.info.indices != []]
 
         if "table_name" in keywords:
             if len(keywords["table_name"]) > 20:
@@ -365,26 +365,28 @@ class TdatHeader(basic.BasicHeader):
                 raise TdatFormatError(
                     f"Unrecognized data type {col.dtype} for column {col.name}."
                 )
-            field_line = f"field[{col.name}] = {ctype}"
+            field_line = f"field[{col.info.name}] = {ctype}"
 
-            if col.format is not None:
-                field_line += f":{col.format}"
-            if col.unit is not None:
+            col_info_meta = col.info.meta or {}
+
+            if col.info.format is not None:
+                field_line += f":{col.info.format}"
+            if col.info.unit is not None:
                 field_line += f"_{col.unit:vounit}"
-            if "ucd" in col.meta:
-                field_line += f" [{col.meta['ucd']}]"
-            if "index" in col.meta:
-                field_line += f" ({col.meta['index']})"
-            elif (indices != []) and (col.name == indices[0]):
+            if "ucd" in col_info_meta:
+                field_line += f" [{col_info_meta['ucd']}]"
+            if "index" in col_info_meta:
+                field_line += f" ({col_info_meta['index']})"
+            elif (indices != []) and (col.info.name == indices[0]):
                 field_line += " (key)"
-            elif col.name in indices:
+            elif col.info.name in indices:
                 field_line += " (index)"
-            if col.description is not None:
-                field_line += f" // {col.description}"
-            elif "comment" in col.meta:
+            if col.info.description is not None:
+                field_line += f" // {col.info.description}"
+            elif "comment" in col_info_meta:
                 field_line += " //"
-            if "comment" in col.meta:
-                field_line += f" // {col.meta['comment']}"
+            if "comment" in col_info_meta:
+                field_line += f" // {col_info_meta['comment']}"
             lines.append(field_line)
 
         if len(keywords) != 0:
@@ -402,7 +404,7 @@ class TdatHeader(basic.BasicHeader):
         lines.append("#")
         lines.append("# Data Format Specification")
         lines.append("#")
-        lines.append(f"line[1] = {' '.join([col.name for col in self.cols])}")
+        lines.append(f"line[1] = {' '.join([col.info.name for col in self.cols])}")
         lines.append("#")
 ```
I tried pushing a commit for this directly but it didn't work.

taldcroft commented:
@d-giles - is this ready for final review now or are you still working on it? I see you have pushed several more commits recently.

d-giles commented:
@taldcroft - I think I've addressed most of the lingering issues and failing tests so it's ready for final review.

taldcroft commented:
@d-giles - I've had a first pass through this, mostly just getting a feel for the format and the code. Overall this looks great, so thanks for the effort in putting it together!!

It is great that TDAT has a reasonably complete requirements document. This gives an opportunity to map every element in the requirements to one or more tests. I understand this can be a bit time-consuming, but I'm not going to be able to scrutinize every line of code and every regex. Instead I'd like to focus on tests and documenting coverage of all of the TDAT requirements (or documenting exceptions).

Overall this shouldn't be too bad since many of them can be covered in a small number of test files / tables.

As a starting example of missing coverage:

_Many of the parameter values (the portion of the parameter definition that comes after the equal sign) can be encased in quotes. Double quotes, single quotes, and back quotes are all recognized, but each pair of quotes must match in type._

From what I can see only single quote is currently supported, meaning that when a double quote or back quote is present then they are included in the header value instead of being stripped.

I also note that writing to a TDAT file with missing values and a space delimiter succeeds, but reading back that file fails.

taldcroft commented on astropy/io/ascii/tests/test_tdat.py:
```diff
@@ -0,0 +1,394 @@
+    Write a full-featured table with common types and explicitly check output
+    """
+    t = test_table
+    lines = [
```
Is `lines` here exactly the same as `test_dat`? If not, what are the differences and why?

d-giles replied:
There are a couple differences
* Empty comment lines are dropped (except when denoting a section header like "Table Parameters"). 
* The data type for string field "name" is downsized to char3 from char12 reflecting the maximum size in the column. 
* Extraneous spaces in the data are stripped

taldcroft replied:
Sounds fine.

taldcroft commented on astropy/io/ascii/tests/test_tdat.py:
```diff
@@ -0,0 +1,394 @@
+
+def test_write_read_roundtrip():
+    """
+    Write a full-featured table with all types and see that it round-trips on
```
The test `test_write_full` tests that TDAT -> Table -> TDAT round-trips. 

This test is doing the converse of Table -> TDAT -> Table. That's good, but there are no tests anywhere that the Table itself looks as expected. Are the Table columns all correct (dtype, description, units, etc)?  Is the Table meta correct? Just because it round-trips to the same TDAT file doesn't necessarily imply that the Table value is correct.

An example is that the `'table_description = "Test table"'` header key/value is being parsed as the string `'"Test table"'` with the surrounding double quotes in the string value. By not confirming the values of the Table meta, issues like this can be missed.

taldcroft replied:
You can use the TDAT writer to create a TDAT text table. Then confirm by hand that it matches expectation (e.g. no extra quotes etc). Then that TDAT text should round trip exactly through Table.

esabol commented:
What's needed here to get this merged?

taldcroft commented on astropy/io/ascii/tests/test_write.py:
```diff
@@ -878,16 +878,18 @@ def test_roundtrip_masked(fmt_name_class):
     out = StringIO()
     fast = fmt_name in ascii.core.FAST_CLASSES
+
+    if "tdat" in fmt_name:
```
Why is this not `if fmt_name == "tdat":`? We don't want the block below running for a format name "fatdat".

d-giles replied:
I wasn't aware of any ambiguity with the name and was lazily following the format a couple lines above (873: `if "qdp" in fmt_name:`). I also wasn't sure if "ascii.tdat" needed to be included, but looking at it I think your proposed change is appropriate.

taldcroft commented on astropy/io/ascii/tests/test_write.py:
```diff
@@ -878,16 +878,18 @@ def test_roundtrip_masked(fmt_name_class):
     try:
         ascii.write(t, out, format=fmt_name, fast_writer=fast)
     except ImportError:  # Some failed dependency, skip test
         return
```
Please revert the two changes that remove newlines below.

taldcroft commented:
One question is about `table_name`, which apparently has specific requirements in this format:

table_name = origin_tablename
Required. The supplied table name must either be a recognized system table ("zzgen", "zzext", "zzpar", and "zzrel") or begin with "origin_" where "origin" is the originator of the database. Currently, the only recognized origin is "heasarc"; however, that is easily changed.

Is that in play here or is the intent to just ignore this requirement in the TDAT spec?

taldcroft commented on astropy/table/tests/test_mixin.py:
```diff
@@ -114,11 +114,16 @@ def test_io_ascii_write():
     t = QTable(MIXIN_COLS)
     for fmt in _get_connectors_table():
-        if fmt["Write"] and ".fast_" not in fmt["Format"]:
+        if fmt["Write"] and ".fast" not in fmt["Format"]:
```
Why is the trailing `_` removed?

d-giles replied:
My mistake, I had changed the line in a prior commit to skip if "tdat" or ".fast" were in fmt["Format"] and I manually reverted the line incorrectly.

taldcroft commented on astropy/table/tests/test_mixin.py:
```diff
@@ -114,11 +114,16 @@ def test_io_ascii_write():
+    import warnings
+
     from astropy.io.ascii.connect import _get_connectors_table
+    from astropy.io.ascii.tdat import TdatFormatWarning
```
I think it would be better to special-case this format and set `.meta["table_name"]` in the table. Suppressing these warnings could be masking some other warning since there are a number of ways `TdatFormatWarning` can be raised.

taldcroft commented on astropy/table/table_helpers.py:
```diff
@@ -94,7 +94,7 @@ def complex_table():
         votable = parse(
             get_pkg_data_filename("../io/votable/tests/data/regression.xml"),
-            pedantic=False,
+            verify="ignore",
```
Why this change? It seems really out of scope.

d-giles replied:
I was, at one point, seeing if I could make tests with the complex_table and noticed that the keyword `pedantic` was replaced by `verify` in version 4.0, and deprecated in 5.0. It is out of scope and, since I didn't ultimately incorporate a complex table into any tests, unrelated. I can revert and open a separate issue if you'd like.

taldcroft commented on docs/changes/io.ascii/16780.feature.rst:
```diff
@@ -0,0 +1,3 @@
+Add functionality to read and write to a Table from the TDAT format as part of
+the Unified File Read/Write Interface and tests for basic functionality.
+Fixes issue #15190
```
We don't usually include the fixes issue #, especially for an enhancement like this. Also, no need to mention adding tests, that is always the case!

taldcroft reviewed (commented):
Quick review of a few things that jumped out. One other thing is the What's New, which as mentioned needs to be for 7.1 now.

esabol commented:
> One question is about `table_name`, which apparently has specific requirements in this format:
> 
> table_name = origin_tablename Required. The supplied table name must either be a recognized system table ("zzgen", "zzext", "zzpar", and "zzrel") or begin with "origin_" where "origin" is the originator of the database. Currently, the only recognized origin is "heasarc"; however, that is easily changed.
> 
> Is that in play here or is the intent to just ignore this requirement in the TDAT spec?

For the purposes of astropy, `origin_` should be considered optional. Most of that part of the "spec" can be ignored as it is documenting HEASARC internal usage only. If there is an origin, it probably should match `[a-z][a-z0-9]*` though, and the whole `table_name` value should ideally only contain a single underscore.

esabol commented:
Could someone remove the "Close?" label that the gitbot added to this PR? Thank you!

taldcroft commented:
> Could someone remove the "Close?" label that the gitbot added to this PR? Thank you!

Done.

pllim commented:
Hello. To truly reset stale bot timer, please rebase. Then you get another 5 months.

esabol commented on astropy/io/ascii/tdat.py:
```diff
@@ -61,9 +61,84 @@ class TdatFormatWarning(AstropyWarning):
+    Multiple data delimiters - Multiple delimiters can be specified in the
+        header, e.g. field_delimiter = "|!" would treat both | and !
+        as individual delimiters. Default: "|"
+    Record Delimiters - The record_delimiter can be specified in the header. By
```
`record_delimiter` was deprecated back in the '90s. It was a bad idea, I think, and I doubt it's at all useful. I recommend removing all support for it, but you've already implemented it, so up to you.

`field_delimiter` is also deprecated, but it's potentially useful when dealing with comma-separated values or tab-separated values, so you might as well keep that, imho.

By the way, if the field_delimiter is `|` and some field contains a `|` character, you can escape the field delimiter by putting a backslash in front of it. And `\\` should be interpreted as a single backslash, of course. This is fully supported by our ingest software, and you might want to make sure your code works similarly.

d-giles replied:
I've only included the ability to read TDAT files with the deprecated delimiter keywords, and disallowed them in writing (issuing a warning that they're ignored).

You're saying that `|` could be part of the data in a character column? I will check how that's handled, I anticipate I'll need to make some changes.

esabol replied:
> You're saying that | could be part of the data in a character column?

Yes, it's rare but possible.

esabol replied:
So
```
one field|two field|red\|field|blue\\field|
```
would result in the following values for this row in the database table:

`one field`, `two field`, `red|field`, `blue\field`.

And vice versa if you were to export from the database to TDAT.

d-giles replied:
While deprecated, I want to cover the possibility of ingesting old files. I presume this holds true for all other delimiter characters as well...

esabol replied:
> While deprecated, I want to cover the possibility of ingesting old files.

I think I can definitively state that no files with non-default `record_delimiter`s exist in the whole world, unless you created some to test your PR with. I'm pretty sure I don't have any, and I implemented the software which reads and writes TDAT files for the HEASARC. It was deprecated _very_ early on. That's why I said `record_delimiter` was not useful. But if you want to keep it in, that's fine with me. It just seems like an unnecessary complication that could potentially be difficult to support, but it's your call. Thanks again for you work on this!

esabol replied:
> I presume this holds true for all other delimiter characters as well...

Oh, I just realized you were commenting about `field_delimiter`, huh? Oops! Yes, a backslash can be used to escape any `field_delimiter` character that appears within the value for a field, not just `|`.

d-giles replied:
I did wonder about `record_delimiter`s, but I hadn't gotten to them yet

esabol commented on astropy/io/ascii/tdat.py:
```diff
@@ -376,10 +469,10 @@ def write(self, lines):
                 TdatFormatWarning,
             )
-            lines.append("table_name = astropy_table")
+            lines.append("table_name = origin_astropy")
```
As per my comment on the intent of `origin` in the file format documentation, I suggest changing this back to `astropy_table` throughout.

esabol commented on astropy/io/ascii/tdat.py:
```diff
@@ -388,6 +481,11 @@ def write(self, lines):
+        for kw in ["record_delimiter", "field_delimiter"]:
+            val = keywords.pop(kw, None)
+            if val is not None:
+                warn(TdatFormatWarning(f"Skipping deprecated keyword: {kw}."))
```
Oh, these are deprecated? Well, I guess you can ignore some of my previous comments.

d-giles commented on astropy/io/ascii/tdat.py:
```diff
@@ -162,15 +237,19 @@ def _process_keywords(self, lines):
+
                 value = kmatch.group("value").strip("\"'`")
                 if key == "table_name":
+                    pattern = r"^(zzgen|zztext|zzpar|zzrel|origin_.*)$"
```
Ah I see, I misunderstood. In terms of strictness, do you think even warning the user of an unrecognized origin/table_name is too much?

esabol replied:
> In terms of strictness, do you think even warning the user of an unrecognized origin/table_name is too much?

I'm not sure. I probably wouldn't, but if the Astropy folks think it's a good idea, then go with that, of course.

Thanks for all your work on this PR! 👍

d-giles replied:
On review, I think the fact that ingestion issues its own warning but ultimately continues ingestion makes me think the astropy warning is unnecessary.

The spec doesn't say that `table_name` must begin with a character, or is limited to a single underscore. I'm inclined not to impose restrictions without documentation to point to.

esabol replied:
> The spec doesn't say that table_name must begin with a character, or is limited to a single underscore. I'm inclined not to impose restrictions without documentation to point to.

I wrote the documentation, and I concur. :smile: In practice, the DBMS (PostgreSQL or whatever) will enforce it's own restrictions on what's a legal table name.

esabol reviewed (approved):
Looks good to me! Thank you, @d-giles !

taldcroft commented:
@d-giles - is this ready again for final review?

pllim commented:
> d-giles requested a review from taldcroft 19 hours ago

@taldcroft , I think so. Review away! :)

taldcroft commented:
@d-giles - a few more comments to follow, this is looking pretty close.

- There needs to be an update to various bits of the astropy user docs to indicate this new format. HOWEVER, this will definitely cause problems for #17804, so we need to plan on a follow-up  docs update after #17804 is merged.
- Multi-component units like m/s do not appear to work. The spec document has no advice on units like that and how to write them.
- The examples in the spec doc use `degrees` but this gets written out as `deg`. Is that a problem?

taldcroft commented:
There should be a test of all possible data types for an input file:
```
	       int1     -- a one-byte integer
	       integer1 -- "    "        " (usage discouraged)
	       tinyint  -- "    "        " (usage discouraged)
	       int2     -- a two-byte integer
	       integer2 -- "    "        " (usage discouraged)
	       smallint -- "    "        " (usage discouraged)
	       int4     -- a four-byte integer
	       integer4 -- "     "        " (usage discouraged)
	       integer  -- "     "        " (usage discouraged)
	       float4   -- a four-byte floating point number
	       real     -- a synonym for 'float4' (usage discouraged)
	       float    -- an eight-byte floating point number (usage discouraged)
	       float8   -- "      "         "       "     "
	       char(#)  -- character string, # must be in the range 1 to 2000 (usage discouraged)
	       char#    -- character string, # must be in the range 1 to 2000
```

esabol commented:
@taldcroft wrote:
> * Multi-component units like m/s do not appear to work. The spec document has no advice on units like that and how to write them.

How to write units is well beyond the scope of the TDAT file format, with one exception: Units should not and cannot have spaces in them. Use a period (for multiplication of units) or an underscore if a space is needed, such as an `Solar_Mass`. 👀 @d-giles 

> * The examples in the spec doc use `degrees` but this gets written out as `deg`. Is that a problem?

The HEASARC's TDATs  use `degree`, for the record, but, no, it's not a problem. The TDAT file format is units-agnostic, and people should feel free to write their units however they want to (except for the aforementioned "no spaces" aspect).

taldcroft commented:
> Units should not and cannot have spaces in them. Personally, I would use a period to multiply units ...

Conveniently enough, the CDS format for unit seems to match the requirement and your preference:
```
>>> unit = u.m / u.s**2
>>> f"{unit:cds}"
'm.s-2'
>>> u.Unit("m.s-2", format="cds")
Unit("m / s2")
```
So I would suggest formally adopting and documenting CDS as the standard for unit representation in the astropy TDAT reader/writer. See: https://vizier.unistra.fr/vizier/doc/catstd-3.2.htx.

taldcroft commented:
Currently the TDAT implementation does not correctly round trip a table with compound units and it creates an invalid TDAT file with spaces in the unit. Both these issues need to be addressed before this PR can be merged.
```
>>> t = QTable()
>>> t.meta["table_name"] = "test"
>>> t["a"] = [1, 2, 3] * u.m / u.s**2
>>> t.write("test.tdat", format="ascii.tdat", overwrite=True)
>>> t2 = QTable.read("test.tdat", format="ascii.tdat")
>>> print(t2["a"].unit)
m
```
```
#
# Table Parameters
#
field[a] = float8_m / s2
```

d-giles commented on docs/whatsnew/7.1.rst:
```diff
@@ -33,6 +34,17 @@ It is now possible to remove those when reading a table with
+The TDAT format, specified by https://heasarc.gsfc.nasa.gov/docs/software/dbdocs/tdat.html,
+is now supported by the Unified File Read/Write Interface.
+
+Note that this format is compatible with ``Table`` objects with integer, float,
```
I was running into issues with the format early on, but I think that those issues have since been resolved. I'll add a test reading and writing via QTables

d-giles commented:
> Multi-component units like m/s do not appear to work. The spec document has no advice on units like that and how to write them.
> The examples in the spec doc use degrees but this gets written out as deg. Is that a problem?

> So I would suggest formally adopting and documenting CDS as the standard for unit representation in the astropy TDAT reader/writer. See: https://vizier.unistra.fr/vizier/doc/catstd-3.2.htx.

> I'm well aware of all of the standards/conventions for writings, and I maintain they are unrelated and well beyond the scope of the TDAT file format. It is confounding to me why you are bringing this up here.

To summarize this issue, `astropy` converts recognized units to `astropy.units.Unit` objects when ingested (including changing "degrees" to "deg"), `Table` and `QTable` objects will convert given units if possible when created, or create a custom `Unit` ("fake1/fake2" is treated as a single unit, not multiple). When written, a `Unit` object gets written out using spaces to separate recognized units rather than periods by default. While unspecified in the TDAT documentation, spaces in the unit are not permissible. The default behavior _must_ be set for recognized units otherwise spaces _will_ be the default, CDS appears to be a viable default for the required format.

On read-in: spaces in "field" lines are treated as delimiters between subfields, though this appears to be extraneous as each subfield has its own unique demarcation. A space in the unit (or any other subfield) confuses the parser since it then finds characters which do not meet the patterns of specified subfields. All subfields following the space get lost.

* CDS seems like a good default given that a default must be set for writing `astropy` units. 
* I can set the parser to ignore anything between expected delimiters (spaces and word characters), or I can strip the field line (before description and comments) of all spaces before parsing. I'm inclined towards the latter.

d-giles commented on astropy/io/ascii/tdat.py:
```diff
@@ -0,0 +1,812 @@
+
+    def _process_keywords(self, lines):
+        line_fields = OrderedDict()
+        keywords = getattr(self, "_keywords", OrderedDict())
```
Doesn't seem like it should matter, and tests succeed with `dict`, I'll change to regular `dict` objects

d-giles commented:
> * There needs to be an update to various bits of the astropy user docs to indicate this new format. HOWEVER, this will definitely cause problems for [Refactor and rework documentation on astropy file I/O #17804](https://github.com/astropy/astropy/pull/17804), so we need to plan on a follow-up  docs update after [Refactor and rework documentation on astropy file I/O #17804](https://github.com/astropy/astropy/pull/17804) is merged.

Now that the referenced PR is merged, should I rebase and go ahead and update the user docs?

taldcroft commented:
> > * There needs to be an update to various bits of the astropy user docs to indicate this new format. HOWEVER, this will definitely cause problems for [Refactor and rework documentation on astropy file I/O #17804](https://github.com/astropy/astropy/pull/17804), so we need to plan on a follow-up  docs update after [Refactor and rework documentation on astropy file I/O #17804](https://github.com/astropy/astropy/pull/17804) is merged.
> 
> Now that the referenced PR is merged, should I rebase and go ahead and update the user docs?

Yes please! :smile: There at least three places where this format should be mentioned. (Maybe one day this will be automated, but right now it is manual).
- https://docs.astropy.org/en/latest/io/unified_table.html#built-in-table-readers-writers
- https://docs.astropy.org/en/latest/io/unified_table_text.html
- https://docs.astropy.org/en/latest/io/ascii/index.html#id1

taldcroft commented:
> CDS seems like a good default given that a default must be set for writing astropy units.

Agreed.

> I can set the parser to ignore anything between expected delimiters (spaces and word characters), or I can strip the field line (before description and comments) of all spaces before parsing. I'm inclined towards the latter.

At first glance it seems like the format is sufficiently specified that it should be possible to uniquely extract the string representing the units field. If that is possible, it would be best to either (1) require no spaces (and raise an exception otherwise) or (2) allow spaces and parse the unit field using the astropy default parsing.

I would prefer this over silently ignoring parts of the unit field or stripping spaces. Both of those are likely to surprise users, e.g. if the unit field is `deg s` then both `deg` and `degs` are not expected.

taldcroft commented on docs/io/ascii/index.rst:
```diff
@@ -200,6 +200,7 @@ are compatible with the fast Cython/C engine for reading and writing.
 ``rst``                     Yes      :class:`~astropy.io.ascii.RST`: reStructuredText simple format table
 ``sextractor``                       :class:`~astropy.io.ascii.SExtractor`: SExtractor format table
 ``tab``                     Yes  Yes :class:`~astropy.io.ascii.Tab`: Basic table with tab-separated values
+``tdat``                    Yes      |TDAT|: Transportable Database Aggregate Table format
```
In all four of the references to `TDAT`, the link should follow the pattern and be:
```
:class:`~astropy.io.ascii.Tdat`
```

taldcroft commented on astropy/io/ascii/tdat.py:
```diff
@@ -0,0 +1,815 @@
+            if col.info.format is not None:
+                field_line += f":{col.info.format}"
+            if col.info.unit is not None:
+                if " " in f"{col.info.unit}":
+                    unit_str = f"{col.info.unit:cds}"
+                else:
+                    unit_str = f"{col.info.unit}"
+                field_line += f"_{unit_str}"
```
For consistency and simplicity, always output using CDS unit formatting. Then the `Tdat` class docstring should mention that units are always written in CDS format.
```suggestion
                field_line += f"_{col.info.unit:cds}"
```

taldcroft commented on astropy/io/ascii/tdat.py:
```diff
@@ -0,0 +1,815 @@
+    When writing to the TDAT format, the header will be auto-populated by
+    information in the Table, prioritizing information given in the Table.meta:
+
+    comments : list or string, (optional)
+    Table information which provide context. This information is
+    included in the header preceding all other lines and commented
+    out with #
+
+    keywords : dict, (optional, recommended)
+    Header keywords which will appear in the file as "name=value" lines.
+    Of particular importance are table_name, table_description,
+    and table_document_url.
```
You can format this more nicely like below, so it looks like https://docs.astropy.org/en/stable/io/ascii/read.html#parameters-for-read
```suggestion
    **comments** : list or string, (optional)
      Table information which provide context. This information is
      included in the header preceding all other lines and commented
      out with #

    **keywords** : dict, (optional, recommended)
      Header keywords which will appear in the file as "name=value" lines.
      Of particular importance are table_name, table_description,
      and table_document_url.
```

taldcroft commented on astropy/io/ascii/tdat.py:
```diff
@@ -0,0 +1,815 @@
+    column properties and the recommendations for the TDAT format by HEASARC.
+
+    Example::
+    >>> from astropy.table import Table
```
Blank line.

taldcroft commented on astropy/io/ascii/tdat.py:
```diff
@@ -0,0 +1,815 @@
+    >>> t = Table(names=('reference_id', 'RA', 'Name'),
+    ...           data=[[1, 2, 3], [1.0, 2.0, 3.0], ['c', 'd', 'e']])
+    >>> t.meta['table_name'] = "astropy_table"
+    >>> out = StringIO()
+    >>> t.write(out, format="ascii.tdat")
+    >>> out.getvalue().splitlines()
```
The splitlines output is not readable.
```suggestion
    >>> print(out.getvalue())
```

taldcroft replied:
```suggestion
    >>> t.write(sys.stdout, format="ascii.tdat")
```

taldcroft commented on astropy/io/ascii/tdat.py:
```diff
@@ -0,0 +1,815 @@
+    Including relevant metadata for the table and columns separately
+    is possible with a mixture of attribute assignment and additions to the
+    metadata::
+    >>> from astropy.table import Table
```
Blank line.

taldcroft commented on astropy/io/ascii/tdat.py:
```diff
@@ -0,0 +1,815 @@
+
+    Example::
+    >>> from astropy.table import Table
+    >>> from io import StringIO
```
```suggestion
    >>> import sys
```

taldcroft commented on astropy/io/ascii/tdat.py:
```diff
@@ -0,0 +1,815 @@
+    >>> t.columns['RA'].meta['ucd'] = "pos.eq.ra"
+    >>> t.columns['Name'].description = "The name of the source (if available)"
+    >>> out = StringIO()
+    >>> t.write(out, format="ascii.tdat")
```
Same as above, use `t.write(sys.stdout, format="ascii.tdat")`.

taldcroft commented on astropy/io/ascii/tdat.py:
```diff
@@ -0,0 +1,815 @@
+      table_name example_table
+      table_description Example table
+      table_author Example et al.
+
```
Remove extra blank line here.
```suggestion
```

taldcroft commented on astropy/io/ascii/tdat.py:
```diff
@@ -0,0 +1,815 @@
+
+class Tdat(core.BaseReader):
+    """TDAT format
+    See: https://heasarc.gsfc.nasa.gov/docs/software/dbdocs/tdat.html"
```
```suggestion
    See: https://heasarc.gsfc.nasa.gov/docs/software/dbdocs/tdat.html
```

taldcroft commented on astropy/io/ascii/tdat.py:
```diff
@@ -0,0 +1,815 @@
+    """TDAT format
+    See: https://heasarc.gsfc.nasa.gov/docs/software/dbdocs/tdat.html"
+
+    Example::
+      <HEADER>
```
```suggestion
    Example::
    
      <HEADER>
```

taldcroft commented on astropy/io/ascii/tdat.py:
```diff
@@ -0,0 +1,815 @@
+
+
+class Tdat(core.BaseReader):
+    """TDAT format
```
```suggestion
    """TDAT format
    
```

taldcroft replied:
Overall for the docstring here, you should build the docs from source and make sure they look good before pushing docs commits to GitHub. Sphinx/RST is really fiddly and I never get it right the first time, so building locally prevents burning CI hours.

https://docs.astropy.org/en/latest/development/docguide.html#building-the-documentation-from-source

taldcroft reviewed (commented):
Almost there!

taldcroft commented:
The TDAT format only allows up to 4-byte integers. The current code will write output that exceeds this limit with no warnings.
```
In [32]: t = Table([[-2**60, 2**60]])

In [33]: t.write(sys.stdout, format='ascii.tdat')
...
<HEADER>
table_name = astropy_table
#
# Table Parameters
#
field[col0] = int4
#
# Data Format Specification
#
line[1] = col0
#
<DATA>
-1152921504606846976|
1152921504606846976|
<END>
```
Should this be handled with a warning or exception?

esabol commented:
> The TDAT format only allows up to 4-byte integers. The current code will write output that exceeds this limit with no warnings.

There just hasn't been a compelling reason to add support for 8-byte integers. There have probably been only a few tables in the past almost 30 years that needed them, and it's been easier to just use char fields for these rare cases, even if it's storage-inefficient.

That said, I can _easily_ add the `int8` field type to the TDAT format documentation if that makes it easier for you to implement things in astropy.

d-giles commented:
If I'm not mistaken part of the prior comment is from @taldcroft via an edit rather than @esabol, is that right?
> The TDAT format only allows up to 4-byte integers. The current code will write output that exceeds this limit with no warnings.
> One thing I just noticed which relates is that a field which is typed in TDAT as `int4` is actually read as a numpy `np.int64` array. That doesn't seem expected since `int2` gives `np.int16` and `int1` => `np.int8`.

I had done this originally to match the python default integer on read. On write, if a column had an integer dtype that wasn't directly specified was given, the writer would default to `int4`, but as you found, it didn't actually enforce that the values were compatible. I've changed this so that it is more consistent internally (`int4` -> `np.int32` on read, see below for write)

> About the question of "when it's actually needed", the general behavior is that an `np.int64` array would "really need" `int8`. In other words, don't ask questions about the values, just use the declared array dtype to determine the requirement in the output file.


I've changed the output to try and downsize the dtype to `int4` if it's not one of `np.int8`, `np.int16`, or `np.int32` if possible. It'll throw an error if any of the values in the column are too large for `int4`.

esabol commented:
> If I'm not mistaken part of the prior comment is from @taldcroft via an edit rather than @esabol, is that right?

Yes. @taldcroft edited my post instead of replying in a separate post. I've made the same mistake on projects myself on occasion. Over half of my post was removed. I've re-edited the post to fix it as best as I can.

taldcroft commented on astropy/io/ascii/tdat.py:
```diff
@@ -0,0 +1,853 @@
+        self._set_col_formats()
+        for col in self.cols:
+            if np.issubdtype(col.dtype, np.integer):
+                if np.any(abs(col.data) > 2147483647):
```
The allowed range for int32 is `-2**31` to `2**31 - 1`, so this test can't use `abs(...)`.

taldcroft reviewed (commented):
The latest updates look good. Just one tiny fix and this should be good for merging!

taldcroft commented:
Sorry about the accidental edit, I'm not sure how that happened.

taldcroft reviewed (approved):
This looks great, thanks for the tremendous effort @d-giles! Also thanks to @esabol for the review and inputs.

esabol commented:
Thanks, @d-giles and @taldcroft ! This took a lot more effort than I thought it would when I opened the issue, and I really appreciate @d-giles hanging in there and seeing this through. I'm happy to see this merged. Cheers!

pllim commented:
I think this needs additional patch to work in s390x . I see failure from what I think is this PR in another PR here: https://github.com/astropy/astropy/actions/runs/14375617614/job/40307218693?pr=17995

I just kicked off a run on main here: https://github.com/astropy/astropy/actions/runs/14379962816

pllim commented:
p.s. For the future, any PR with significant change, especially completely new feature could benefit from both "Extra CI" and "Build all wheels" labels.

Thanks, all!

dhomeier commented on astropy/io/ascii/tdat.py:
```diff
@@ -0,0 +1,853 @@
+                ctype = "int4"
+            elif "float" in str(col.info.dtype):
+                ctype = "float8"
+            elif "<U" in str(col.info.dtype):
```
Is little-endian mandated by the format, or could we also allow `>U` (or `=U`?) as dtype here? This is apparently the cause of #17997.

esabol replied:
The format is ASCII, so it should be unaffected by endianness.

dhomeier replied:
That's more or less what I expected, too. So why specifically filtering for little-endian types here? Note that I haven't delved far into the context yet.

mhvk commented on astropy/io/ascii/tests/test_tdat.py:
```diff
@@ -0,0 +1,817 @@
+    ):
+        test_table.write(out, format="ascii.tdat", delimiter=",")
+
+    test_table.meta["keywords"]["field_delimiter"] = ","
```
This modification of the table without making a copy caused problems with the double-test run. See #18025 for a fix. What I don't understand is how the tests could have passed before...
